### ISSUE TITLE
barchart: allow to add multiple bars from different datasets

### ISSUE THREAD

**Issue by karthago1** [tui-rs-revival/ratatui#294]
## Problem
currently the barchart allows to add only one data set. this is useful to show for example revenue chart of a company every month. but if we want to do a comparison of different companies, then we need to be able to add multiple datasets. (every dataset belongs to a different company). example:

![barchart](https://github.com/tui-rs-revival/ratatui/assets/15311024/9e3cf1d1-df5f-42df-947d-52b5d3ce7930)

## Solution
here is a possible implementation #288.:

the data function is modified from
`pub fn data(mut self, data: &'a [(&'a str, u64)]) -> BarChart<'a> `
to: 
`pub fn data(mut self, data: impl Into<Group<'a>>) -> BarChart<'a>`

- a Group and Bar structs where introduced. (Group has a vector of bars)
- group_gap: set the gap between each bar group

**Comment by joshka** [comment]
Thanks for making this issue @karthago1. Moving some of the discussion from the PR to here.

In #288 I wrote:

---
I like the idea and direction - perhaps consider documenting the user's perspective a little before continuing into the code. (It's important to tell us what you want to do, so that when it's complete, it's obvious).

I notice that this adds a new "Group" feature, but "Group" only appears in the API once. How would you design the API so that it made this feature more explicitly obvious?

I think it's probably possible to keep the single `data()` function, and expect the user to call it multiple times rather than adding a new `add_data()` function which is confusingly similar.

--- 
In #288 @karthago1 wrote:

> I think it's probably possible to keep the single data() function, and expect the user to call it multiple times rather than adding a new add_data() function which is confusingly similar.

true. the problem with `data()`, that it passes always one dataset with the labels. the labels should be the same for all datasets. so if the user calls `data()` twice. the first time I can store the labels. and the second call should I ignore them?

possible Solutions:
1. remove `data()` and rename `add_data()` to `data()`. this is not backwards compatible solution. labels can be passed by the function `labels()` instead of the old `data()`
2. maybe I can modify `data()` to pass all datasets at once and stay backwards compatible too. I will check this possibility
3. keep the same `data()` and add new function `add_data()` which adds only one data set. (current solution)

the following 2 code sections produces the same result:
```
BarChart::default()
        .data(&[("Mar", 9), ("Apr", 12), ("May", 5), ("Jun", 8)])
        .add_data(&[1, 2, 3, 4])
        .group_gap(4);
```

```
BarChart::default()
        .add_data(&[9, 12, 5, 8])
        .add_data(&[1, 2, 3, 4])
        .group_gap(4)
        .labels(&["Mars", "Apr", "May", "Jun"]);
```

**Comment by joshka** [comment]
I think it might be reasonable to look at this as each call to .data defines the data for that group, each value has a label, and each group can also have a label. I can think of a few arrangements where it's useful to make the labels in one group different, as well as making the number of items different. For that reason it makes sense to me to provide each group at a time rather than data for the first elements of each group, then the second element etc.

To be backwards compatible, you could make data take something that implements grouped data:
```rust
fn data(group: Into<Group>)
...
struct Group {
  label: Option<Line>,
  data: &[Line, u64)],
}

impl From<&[Line, u64]> for Group {
  fn from(value &[Line, u64]) -> Group {
    Group {
      label: None,
      data: value,
    }
  }
}
```

So your example could become something like:
```rust
BarChart::default()
    .labels(&["Company A", "Company B"]);
    .data(Group::new("Mar", &[9, 1)])
    .data(Group::new("Apr", &[12, 2])
    .data(Group::new("May", &[5, 3])
    .data(Group::with_labels("Jun", &[("Company X", 8), ("Company Y", 4)])
    .group_gap(4);
```

Not sure if Line / Span / Text is more appropriate for each label, but using that rather than just a string allows you to style the text directly rather than having to go via the widget. (e.g. "Company A".blue())

Perhaps we could even have a `Bar` as well, which has title, style, value? Then Group is a collection of Bars?

**Comment by karthago1** [comment]
this is a great idea. this will make the implementation easier. but this will make it a little bit harder for the user, since 90 % of the use cases will have a dataset which contains values belonging to different Groups and the user will have to map the data. but I think this is worth it.

I really like your suggestion, I will implement it and maybe you can review it again.

there is another small problem with the bar labels. for my use case the labels can be bigger than the bar width and I need them to overlap with the other bars, if it is necessary. In my private branch I implemented this already. but I will post some images describing the problem and maybe you can check the user API again. maybe you will more great ideas

**Comment by karthago1** [comment]
just to be sure. in your example:
`.data(Group::with_labels("Jun", &[("Company X", 8), ("Company Y", 4)])`

`Company X`, and `Company Y` should be shown on the bar right? so I could call for example:
`.data(Group::with_labels("Jun", &[("8 Mil", 8), ("4 Mil", 4)])`  8 and 4 Mil will be shown on the bar.

**Comment by joshka** [comment]
>this is a great idea. this will make the implementation easier. but this will make it a little bit harder for the user, since 90 % of the use cases will have a dataset which contains values belonging to different Groups and the user will have to map the data. but I think this is worth it.

I think this is one of those cases where both ways make sense, so you can't please both. I'd have to see non synthetic examples of why people are using this one way or another to understand. That said, I think the way I've suggested is the more flexible approach (please correct me on that if I'm wrong though).

>Company X, and Company Y should be shown on the bar right? so I could call for example:
Probably under each bar (while the group label sits under the entire group)

> there is another small problem with the bar labels. for my use case the labels can be bigger than the bar width and I need them to overlap with the other bars, if it is necessary. In my private branch I implemented this already. but I will post some images describing the problem and maybe you can check the user API again. maybe you will have more great ideas

Yeah, I saw that problem too when playing with it. If perhaps we supported a horizontal barchart then long labels might make more sense. I think right now that there's no way to turn off bar labels and only show them in the legend. So perhaps spec out those requirements a little. E.g.:

Each bar can have a label for the value which can be generated based on the value by the `format` function. If the resulting string is too wide it is ... (truncated, not displayed, wrapped, ?). It can be styled with the default style being...

Each bar can have a label that identifies the contents, it can be styled... with the default style being..., if the contents are too wide for the space then ..., 

Each group can have a label, ...

**Comment by karthago1** [comment]
maybe you can check the API again. I updated the image and description up.

the only missing feature for my use case is to make the values on the bars overlap of each others and that every second value in different Y line. but since the changes are much, I would first try to get this merged and afterward I will deal with the missing feature.

also some unit tests can be added to improve the code coverage. but lets first approve the current API

**Comment by karthago1** [comment]
fixed by PR  #288
